crypto/mlkem/mlkem768.c

```diff
@@ -0,0 +1,1184 @@
+    uint64_t product = (uint64_t)shifted * kBarrettMultiplier;
+    uint32_t quotient = (uint32_t)(product >> kBarrettShift);
+    uint32_t remainder = shifted - quotient * kPrime;
+
+    /*
+     * Adjust the quotient to round correctly:
+     *   0 <= remainder <= kHalfPrime round to 0
+     *   kHalfPrime < remainder <= kPrime + kHalfPrime round to 1
+     *   kPrime + kHalfPrime < remainder < 2 * kPrime round to 2
+     */
+    assert(remainder < 2u * kPrime);
+    quotient += 1 & constant_time_lt_32(kHalfPrime, remainder);
+    quotient += 1 & constant_time_lt_32(kPrime + kHalfPrime, remainder);
+    return quotient & ((1 << bits) - 1);
+}
+
+/*
+ * Decompresses |x| by using an equi-distant representative. The formula is
+ * round(kPrime/2^|bits|*x). Note that 2^|bits| being the divisor allows us to
+ * implement this logic using only bit operations.
+ */
+static uint16_t decompress(uint16_t x, int bits)
+{
+    uint32_t product = (uint32_t)x * kPrime;
+    uint32_t power = 1 << bits;
+    /* This is |product| % power, since |power| is a power of 2. */
+    uint32_t remainder = product & (power - 1);
+    /* This is |product| / power, since |power| is a power of 2. */
+    uint32_t lower = product >> bits;
+
+    /*
+     * The rounding logic works since the first half of numbers mod |power| have a
+     * 0 as first bit, and the second half has a 1 as first bit, since |power| is
+     * a power of 2. As a 12 bit number, |remainder| is always positive, so we
+     * will shift in 0s for a right shift.
+     */
+    return lower + (remainder >> (bits - 1));
+}
+
+static void scalar_compress(scalar *s, int bits)
+{
+    int i;
+
+    for (i = 0; i < DEGREE; i++)
+        s->c[i] = compress(s->c[i], bits);
+}
+
+static void scalar_decompress(scalar *s, int bits)
+{
+    int i;
+
+    for (i = 0; i < DEGREE; i++)
+        s->c[i] = decompress(s->c[i], bits);
+}
+
+static void vector_compress(vector *a, int bits)
+{
+    int i;
+
+    for (i = 0; i < RANK768; i++)
+        scalar_compress(&a->v[i], bits);
+}
+
+static void vector_decompress(vector *a, int bits)
+{
+    int i;
+
+    for (i = 0; i < RANK768; i++)
+        scalar_decompress(&a->v[i], bits);
+}
+
+static
+public_key_RANK768 *public_key_768_from_external(const ossl_mlkem768_public_key *external)
+{
+    return (struct public_key_RANK768 *)external;
+}
+
+static
+private_key_RANK768 *private_key_768_from_external(const ossl_mlkem768_private_key *external)
+{
+    return (struct private_key_RANK768 *)external;
+}
+
+static int mlkem_marshal_public_key(uint8_t *out,
+                                    const struct public_key_RANK768 *pub)
+{
+    /*
+     * replace CBB logic with straight copy to out and memcpy of rho at tail end
+     * TODO(ML-KEM): Check this is OK to protect incorrect buffer(sizes) passed
+     * possibly use WPACKET?
+     */
+    vector_encode(out, &pub->t, kLog2Prime);
+    memcpy(out + encoded_vector_size(RANK768), pub->rho, sizeof(pub->rho));
```

Comment:
I wouldn't use WPACKET here. Just ensure that the allocated output buffer is large enough and perhaps use assert() to check? This function should really return void.

Reply:
How to check the allocated length of the out buffer?